packages/go_router/lib/src/router.dart

```diff
@@ -36,9 +36,7 @@ import 'typedefs.dart';
 ///  * [GoRoute], which provides APIs to define the routing table.
 ///  * [examples](https://github.com/flutter/packages/tree/main/packages/go_router/example),
 ///    which contains examples for different routing scenarios.
-class GoRouter extends ChangeNotifier
-    with NavigatorObserver
```

Comment:
@chunhtai 

Removing `with  NavigatorObserver` is a pretty big breaking change, isn't it?

I think it must at least be clearly stated in the CHANGELOG, but since it was not there, so I was a little confused:
https://pub.dev/packages/go_router/changelog#512

How about this kind of migration guide?

```dart
extension GoRouterX on GoRouter {
  NavigatorState? get navigator => routerDelegate.navigatorKey.currentState;
}
```

Reply:
ah shoot, yes you are right. I totally missed this one. I will prepare a migration guide

Reply:
put up a pr to clean up the change log and add migration guide https://github.com/flutter/packages/pull/2781

Reply:
@chunhtai The migration guide looks good to me.

@mono0926 I'm not sure about the approach you're suggesting. Are we sure that `RouterDelegate` should expose `navigatorKey`? I'm not sure we want users depending on RouterDelegate to access the router's state or configuration.

Reply:
Using the navigatorKey is also consistent with how ShellRoute works, if you want access to a ShellRoute's Navigator state you provide a key to ShellRoute.

It would be good to have some documentation on how to push and pop pageless routes using navigatorKeys with GoRouter or ShellRoute.

Reply:
@chunhtai @johnpryan 

Thanks for the quick response 🙏 

#2781 looks good 👍 

>> How about this kind of migration guide?
>>
>> ```dart
>> extension GoRouterX on GoRouter {
>>   NavigatorState? get navigator => routerDelegate.navigatorKey.currentState;
>> }
>> ```

> I'm not sure about the approach you're suggesting. Are we sure that RouterDelegate should expose navigatorKey? I'm not sure we want users depending on RouterDelegate to access the router's state or configuration.

I tried accessing `routerDelegate.navigatorKey.currentState`, while wondering if this was the right approach or not, but it just worked with my existing code. I was just curious if there were any problems with this approach.
If there are no problems, I thought this approach would require minimal rewriting of the existing code where it is used.

> If you are accessing the NavigatorState from GoRouter.navigator. Consider passing in a navigator key to [GoRouter constructor](https://pub.dev/documentation/go_router/latest/go_router/GoRouter/GoRouter.html) and accessing the navigator state through the global key directly.
> 
> https://docs.google.com/document/d/1HMUZA-lgeGFZ5JyU4g5jOFhbYAbk9ysfz9Ok1tGvqRw/edit?resourcekey=0-7otvjyBeefkQ9xQP9S2-9w

I thought of that approach and thought it might be a more proper approach, but since it would require rewriting other parts of the code, I decided that it might be reasonable to use the extension method as mentioned above for now. 
However, I was not sure if that was the right approach.

A similar alternative solution is the following (as you can see, it violates the lint rule).

```dart
extension GoRouterX on GoRouter {
  // ignore: invalid_use_of_visible_for_testing_member
  NavigatorState? get navigator => routeConfiguration.navigatorKey.currentState;
}
```

Is there a simple and future-proof way to access [GoRouter's navigatorKey](https://github.com/flutter/packages/blob/1a0086d7c3cf1cecbf7fb3359f5698ffc041538e/packages/go_router/lib/src/router.dart#L63) without passing `navigatorKey` from the outside?

Reply:
yes, once we are confident to open up the GoRouterDelegate as a public API, the `routeConfiguration.navigatorKey.currentState;` would be a valid use case.

Right now it is a mess that GoRouter exposes `GoRouterDelegate` but `GoRouterDelegate` class is package private. The reason I haven't complete open up is because there is some more clean up i want to do before we do that.

Reply:
@chunhtai 

Thanks 👍